### ISSUE TITLE
Why is the equal sign escaped in args to quote()

### ISSUE THREAD

**Issue by drok** [ljharb/shell-quote#11]
When converting an array or args into a string that might be saved to a file as a shell script, equal signs are escaped with a backslash. This is not necessary and makes the result script file harder to read. The superfluous escape is not incorrect, but unnecessary, and ugly.

I wonder why this escape is done, what purpose does it serve? Here's an example:
```
quote(["make", "CFLAGS=-DRELEASE"])
```

I expected the command line to be: `make CFLAGS=-DRELEASE`.

Instead, I got: `make CFLAGS\=-DRELEASE` (the `\` is not necessary)

This behaviour seem to be [implemented in quote.ts](https://github.com/ljharb/shell-quote/blob/fdeb0fd102ecdd4ff227a978532598c462312347/quote.js#L14C3-L14C86):

```
return String(s).replace(/([A-Za-z]:)?([#!"$&'()*,:;<=>?@[\\\]^`{|}])/g, '$1\\$2');
```

I see on this list that other characters which have no special meaning to the shell are also quoted: '@', '^', '{' '}'

I would like to note from the [manpage of bash](https://www.man7.org/linux/man-pages/man1/bash.1.html):

> Note that unlike the metacharacters ( and ), { and } are reserved words and must occur where a reserved word is permitted to be recognized.

This means the curly brackets are not [metacharacters](https://www.man7.org/linux/man-pages/man1/bash.1.html#DEFINITIONS) ("*A character that, when unquoted, separates words.  One of
the following*: **|  & ; ( ) < > space tab newline**"), they need not be escaped. Example:

```
$ echo {this-is-fine}
{this-is-fine}
```

Thanks!

**Comment by ljharb** [comment]
This package isn't unique to bash; there's many other shells we might be used in - zsh, ksh, sh, dash, to name a few.

Given that the purpose an escaping library like this is correctness, and not readability of the output, I'm inclined to leave it as-is, since "too much escaping" is a trivial problem but breakage is a very expensive one.

**Comment by drok** [comment]
So is there a shell where it would be incorrect to not quote the equal sign?

The problem with "too much escaping" is that more is not necessarily better.

For example, the simple implementation at [quote.js#L8-9](https://github.com/ljharb/shell-quote/blob/fdeb0fd102ecdd4ff227a978532598c462312347/quote.js#L8-9) is wrong.

In that case, a single-quoted string is created with backslashes used to escape single quotes and backslashes. This violates [POSIX.1-2017 2.2.2](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_02_02):
> A single-quote cannot occur within single-quotes.

This case cannot occur due to the `!(/'/).test(s)` condition, but the replace pattern does double up the backspaces, and it shouldn't. A user supplied 3-char string `\ \` becomes a 5-char string `\\ \\` after this treatment, because backslashes inside single quoted strings are not escape characters, ie, they don't need escaping, and escaping them breaks the string:
```shell
$ echo '\ \'
\ \
```

I am not focused on correctness; I'd be happy with good looking equal signs. 😉
Correctness can be someone else's battle.

**Comment by ljharb** [comment]
Correctness should be everyone’s battle; it’s all that matters.

If you can provide some test cases that over-escaping breaks, I’ll be happy to fix them.

**Comment by drok** [comment]
Test case:
> user supplied 3-char string `\ \`

**Comment by ljharb** [comment]
The output is meant to be used in double quotes, I think? `echo "\\ \\"` works fine.

**Comment by drok** [comment]
The problem is that `quote(['\\ \\'])`, ie, the 3-char input `5c 20 5c`, returns `'\\ \\'`, ie, the 7-char sequence `27 5c 5c 20 5c 5c 27` when it should return `'\ \'`, ie, the 5-char sequence `27 5c 20 5c 27`

**Comment by drok** [comment]
For this case, the output `"\\ \\"`, ie, the 7-char sequence `22 5c 5c 20 5c 5c 22` would also be correct, but less readable (a 3-char input represented in 7-char is 4 extra characters)

**Comment by ljharb** [comment]
right - but the output of this library needs to work in double quotes, and `echo "\ \"` doesn't work.

Can you elaborate on why you want readability here?

**Comment by drok** [comment]
> why you want readability here

Eyeballs matter. Readability is useful when the output of the library is used in human-visible text like a blog, an email or a book, for instance by cut-and-pasting into a word processor.

Also, when a teacher wrote a command line on a blackboard, he would write it as `echo '\ \'` and not as `echo "\\ \\"`. If the command line had any complexity, single quotes would be the tool of choice, not double quotes. A student or uninitiated reader would take the teacher's writing as authoritative over a textbook that used the more obfuscated quoting. The teacher shouldn't have to explain that some poorly written textbooks **unnecessarily** give the longer notation.

I use this lib to output diagnostics into a log file. Ie, when things go wrong, frustrated human eyeballs will be looking at my diagnostic output and scrutinize every byte. I don't want to add to the frustration by making that user **unnecesarily** interpret the more obfuscated, double-quoted notation.

If the output were only ever intended to be seen by a machine, it wouldn't matter how obfuscated it is.

**Comment by ljharb** [comment]
I agree that readability is very important in a general sense.

If it's for human eyeballs and not a shell parser, why run it through this library at all?

**Comment by drok** [comment]
> I use this lib to output diagnostics into a log file

> I use this lib to output diagnostics into a log file. Ie, when things go wrong, frustrated human eyeballs will be looking at my diagnostic output

**Comment by ljharb** [comment]
Yes, I read that, but I'm still unclear why the output of this would end up in the log file. Either you're using shell commands to put it into the log file - where the shell will remove the escaping for you - or you're not, and you can put the unescaped output directly into the log file.

(heads up that when someone asks a clarifying question, it's very rude to just repeat the things you said. if they didn't understand the first time, then you need to phrase it differently, not repeat it, for them to understand it)

**Comment by drok** [comment]
> but I'm still unclear why the output of this would end up in the log file

The application spawns some subprocess with arguments given as an array of strings, similar to `execve(executable_name, argv[])`. The argv[] array in some way.

The spawning of this process is logged, and it needs to be in a format that:
1. should be as easy as possible for the user to understand, **unambiguously**.
2. should be able to be cut and pasted into a terminal window to correctly execute the same command, by first removing the escaping, then calling execve with exactly the same argv array that my application did.

Putting an unescaped command line would not allow any of these two objectives. Without the correct quoting it would be ambiguous, and could not be executed without manual tinkering.

My apologies for the need to repeat. It looked as if you had not read my entire earlier response, and asked a question that had been pre-emptively answered.

**Comment by ljharb** [comment]
ahh ok, thanks. so you're saying that the humans aren't looking at the log file itself, they're looking at the log of the "command that logs to the file", which has commands that need some escaping.

How would this user "remove the escaping"? You can't generally just delete all backslashes to accomplish that.

**Comment by drok** [comment]
> How would this user "remove the escaping"? You can't generally just delete all backslashes to accomplish that.

The end user might cut and paste the logfile to stackoverflow, and hope for help from some pedantic nerd. Therefore, it must be correct, not some approximation of what was executed, to pass the scrutiny of someone skilled. Any missing or misplaced backslash will result in *Ahhh... there's your problem* - a well intentioned, but a wild goose chase all the same.

The end user can also attempt to re-run the command, possibly by slightly editing it, maybe to add something or remove something. Perhaps a trial-and-error process to find the exact offending argument, so they can correct the input to the application, devise some workaround, or make an informed report to the application developer. As that developer, I'm much more likely to give a swift, accurate solution when the bug reporter has already done some work to narrow down the root cause.

Just as you say, unquoting is not as simple as deleting some backslashes. On the flip side, in order to produce a working command line from an incorrect or unreadable approximation, it's not enough to add some backslashes or sprinkle some quotes. The user should not need to fix up a defective command line before cutting and pasting it into a terminal.

The end user could very well be a pedantic nerd with the command line surgery skills needed to fully diagnose and remedy the issue. Even then, I need to present in the diagnostic logfile information that is accurate and readable, to help them help themselves.

**Comment by ljharb** [comment]
I totally agree that for your use case, you need something that can be copypasted verbatim and run on the CLI.

Certainly, if there's any inputs we accept that when outputted and executed in double quotes, something unexpected happens, we have a bug, and that should be fixed.

If all you want is to remove some of the escapes, it seems like wrapping (not forking) this library and doing so with its output would be a pretty maintainable approach?

**Comment by soletan** [comment]
> Correctness should be everyone’s battle; it’s all that matters.
> 
> If you can provide some test cases that over-escaping breaks, I’ll be happy to fix them.

Looks like shell-quote is causing an [issue](https://github.com/open-cli-tools/concurrently/issues/487) in [concurrently](https://github.com/open-cli-tools/concurrently/blob/3dc909a9618f92cb8dae2fa62538af9db08a377e/src/command-parser/expand-arguments.ts#L29). 

Nothing's wrong with correctness, but correctness is hard too proof and shell-quote seems to more like breaking a fly on the wheel as it is rather scarce on distinguishing when to escape what. Don't get me wrong for this might sound not the way I mean it, but after checking the readme rather briefly I'm wondering if shell-quote is even meant to support cmd.exe on windows at all. Is it?

If it is, escaping at least `@` this generically seems to be wrong as it is passed by cmd.exe into the invoked command e.g. when running it like this:

```cmd
C:\Users\cepharum\PhpstormProjects\UID\Logistics\WebClient\Source\UiNew>cmd /s /c node log.js --tags \@foo
[
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\Users\\cepharum\\PhpstormProjects\\UID\\Logistics\\WebClient\\Source\\UiNew\\log.js',
  '--tags',
  '\\@foo'
]
```

with a file **log.js** like this:

```javascript
console.dir(process.argv);
```

**Comment by soletan** [comment]
Looks like using single quotes in cmd.exe isn't working either:

```cmd
C:\Users\cepharum\PhpstormProjects\UID\Logistics\WebClient\Source\UiNew>cmd /s /c @node log.js --tags \@foo
[
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\Users\\cepharum\\PhpstormProjects\\UID\\Logistics\\WebClient\\Source\\UiNew\\log.js',
  '--tags',
  '\\@foo'
]

C:\Users\cepharum\PhpstormProjects\UID\Logistics\WebClient\Source\UiNew>cmd /s /c "@node log.js --tags \@foo"
[
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\Users\\cepharum\\PhpstormProjects\\UID\\Logistics\\WebClient\\Source\\UiNew\\log.js',
  '--tags',
  '\\@foo'
]

C:\Users\cepharum\PhpstormProjects\UID\Logistics\WebClient\Source\UiNew>cmd /s /c \@node log.js --tags \@foo
Der Befehl "\@node" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.

C:\Users\cepharum\PhpstormProjects\UID\Logistics\WebClient\Source\UiNew>cmd /s /c '@node log.js --tags \@foo'
Der Befehl "'@node" ist entweder falsch geschrieben oder
konnte nicht gefunden werden.
```

Pardon the German output but the errors are indicating that 

- escaping the @ even at the beginning of the line is wrong where it is usually used to suppress the invoked command from appear on screen when used in a batch file - without escaping cmd is simply ignoring it
- [using single quotes](https://github.com/ljharb/shell-quote/blob/main/quote.js#L9) seems to be wrong there, too

**Comment by ljharb** [comment]
@soletan it does indeed seem to support windows paths (based on PRs that predate my maintainership).

If there's a bug, then a PR with a reduced test case would be most helpful, with or without a fix - it's difficult for me to test on Windows but if there's a pre-existing PR then I can fire up a VM and check it.

**Comment by soletan** [comment]
Thanks for the confirmation.

Since we have an existing approach to escaping special shell characters in one of our own tools which I failed to thoroughly test myself, I might spend some more time on improving on the testing part there and let you know of any results or even provide a revision.

**Comment by ljharb** [comment]
Thanks, that'd be great!

**Comment by soletan** [comment]
Okay, I've spent the weekend implementing a test trying options for escaping characters with and without wrapping them in single or double quotes. The resulting tool is generating configurations for tested shells. I've used this tool to create configurations for Windows' CMD and PowerShell. In addition, I've created a docker setup running the tool against several shells supported by Debian. As a result, there are [about 10 different configuration files](https://gitlab.com/cepharum-foss/quoting-db/-/tree/main/packages/library/db) for as many different shells.

For testing the results, I've also implemented a slight method for quoting a single argument based on a selected configuration. As this is already working, I successfully integrated this approach with our tool I was mentioning before, too.

To conclude, tests reveal a rather diverse situation when it comes to escaping and quoting characters for use with different shells. However, I was struggling to see a way to integrate with your package
- for your package seems to provide a combination of a parser and a quoter with my test results providing a more sophisticated solution for the quoter part, only, and
- for your quoter part is working synchronously while having a maintainable support for different kinds of shells might require a more dynamic approach so that asynchronously fetching a used shell's configuration has become part of my apprach.

I hope you don't mind that I've decided to provide the results as [a dedicated package on npm](https://www.npmjs.com/package/@cepharum/quoting-db) for now. Maybe, you see a way to integrate the results with your package so that other tools such as concurrently can benefit from it.

**Comment by ljharb** [comment]
I guess I’m a bit confused. What i was hoping for was a set of test inputs and outputs, and then results for running those outputs on lots of shells.

I don’t see a value in minmal shell-specific quoting; what I’d want is this package to quote the *union* of what shells need, so the output can be used everywhere.

**Comment by soletan** [comment]
Okay, but there is no single union option. E.g., CMD requires ^ for escaping instead of \. As mentioned before, CMD is requiring double quotes when arguments contain whitespace. It doesn't support to escape them without quotes. Powershell OTOH is benefitting more from single quotes when necessary. In addition, it is requiring single quote inside of single quoted argument to be doubled for escaping. Using backslash does not work there, too. Both Windows shells cause issues when using escape sequence on too many characters for they just keep the backslash when they don't deem it necessary. And Linux shells have a similar support for escaping, but it ain't equivalent.

So, I'm happy to see that one-for-all approach that's also correct in all possible scenarios and capable of recognizing whether some argument is safe to be passed eventually. Just have a \n in your arguments and most shells have issues with properly escaping/quoting that. In best case scenario, the \n just gets lost.

**Comment by ljharb** [comment]
i guess it'd be fine to pivot based on OS - but shell detection is something i'd rather avoid.

**Comment by soletan** [comment]
Hm, maybe there is a misunderstanding here. The package I've created does not require you to pick or even detect a shell. In fact, its default behavior is to use the shell that is Node's default according to the documentation of its `spawn()` function - CMD.exe on Windows and /bin/sh on Linux. It is then using the related configuration that's ready for use. So, it is in fact solely relying on the platform to choose a shell's configuration by default. And there is no need to test or detect anything other than properly processing the provided configuration.

If you don't care for the shell, all you do is ask the helper function I've created for a configuration which is a set of regular expressions and mappings of characters that you can then use to quote arguments and even detect if the provided argument is capable of being safely quoted at all. You can do this all on your own or use another helper function I've implemented.

```javascript
const { getShellConfigurationSync, quote } = require( "@cepharum/quoting-db" );
const configuration = getShellConfigurationSync();

const args = [ "--opt", "some argument with space", "--filter", "!$%><" ];
const quoted = args.map( arg => quote( arg, configuration ) );
```

**Comment by mnardit** [comment]
@ljharb In regards to your comment earlier: "If you can provide some test cases that over-escaping breaks, I'll be happy to fix them" — I recently spent a few hours tracking down a severe real-world breakage caused by this exact over-escaping behavior.

In my case, an AI CLI tool (Claude Code) used shell-quote to sanitize arguments before wrapping them in eval. During a key generation prompt for Tauri, my password contained an exclamation mark (!). shell-quote aggressively escaped it to `\!`.

Because ! has no special meaning inside single quotes in bash, this wasn't just "over-escaping"—it actually mutated the underlying string value. The password was saved to the system with the literal backslash included, totally unbeknownst to me. It silently corrupted my private key signing process.

I wrote a full post-mortem on how this escaping behavior broke a 20-release pipeline here:

https://max.nardit.com/articles/the-invisible-backslash

While "too much escaping" might seem harmless for structural shell commands, when the output of shell-quote is used to pass sensitive string literals (like passwords, secret keys, or base64 hashes) to downstream applications, mutating ! into \! causes irreversible data corruption. This might be a test case worth looking at for future parsing rules!

**Comment by ljharb** [comment]
Thanks! I'd love to see test cases added here.

In this case, it may be that the bug is in Claude Code (which your blog seems to confirm), but the responsibility of this library only ends if we've provided enough configurability for its use case.

I would be stoked to review a PR adding test cases, and/or improving docs - as well as to receive issues covering these use cases.

**Comment by ljharb** [comment]
Thanks everyone for the detailed reports over the years. Let me consolidate where this lands, and what I've changed.

The short version: the escaping is intentional and, as far as I can verify, correct. I checked every disputed character by running `quote`'s output back through `sh`, `bash`, and `zsh`, and they all round-trip to the original value:

- `=` becomes `\=`: prevents a leading word from being read as an assignment (`FOO=bar` as a command is an assignment; `FOO\=bar` is a literal argv[0]).
- `{`, `}`, `,`: brace expansion (`{a,b}` expands in bash/zsh).
- `~` becomes `\~`: tilde expansion.
- `^`, `@`: special under zsh (extendedglob) and csh.
- `!` becomes `\!`: history expansion, and pipeline negation as a leading word.

None of these break when the output is used as intended (verbatim, unquoted, in a POSIX-family shell); the shell removes the extra backslashes and you get the original string back. So this is "more escaping than one specific shell in one specific position needs," which is the deliberate union-of-shells behavior, not a correctness bug. I've added tests pinning it in 1c36f3f.

On the specific concerns raised over the thread:

- @drok, the one genuine correctness bug you found was the single-quoted case, not the `=` one: the 3-char input backslash-space-backslash used to quote to `'\\ \\'` (backslashes doubled inside single quotes, where they are literal). That was real, and it was fixed in #20 (shipped in 1.8.3); it now returns `'\ \'`.

- @soletan, the cmd.exe and PowerShell behavior is out of scope: this library emits POSIX-family shell quoting, and backslash escaping and single quotes simply don't mean the same thing in cmd.exe. There is no single string that is correct for both POSIX shells and cmd.exe, so I'm not going to try to make one; a per-shell package like the one you built is the right home for that. I've documented the POSIX-only scope in the README.

- @mnardit, thank you for the write-up; that was a genuinely nasty failure. I traced it: `quote(['a!b'])` produces `a\!b`, which is correct used verbatim (the shell yields `a!b`). The corruption happened because the already-escaped output was then re-wrapped in single quotes (via `eval '...'`), and inside single quotes a backslash is literal, so `\!` stays `\!`. The escaping wasn't wrong; re-quoting an already-quoted string was. That is a real footgun, so I've documented in 1b36468 that the output must be used verbatim and never wrapped in additional quotes or `eval '...'`.

So the escaping stays, with tests (1c36f3f) and docs (1b36468) added to explain it. If anyone can produce an input that, used verbatim in a POSIX-family shell, does not round-trip, that is a bug and I will fix it immediately, but I was not able to find one. Closing on that basis, and happy to reopen with such a case.
